### ISSUE TITLE
[java] UselessParentheses false positive with conditional operator

### ISSUE THREAD

**Issue by Vampire** [pmd/pmd#1673]
**Affects PMD Version:** *6.11.0*
**Rule:** [UselessParentheses](https://pmd.github.io/latest/pmd_rules_java_codestyle.html#uselessparentheses)

**Code Sample demonstrating the issue:**

```java
throw new IllegalArgumentException(String.format(
        "Unsupported client with class '%s' given",
        (discordClient == null) ? null : discordClient.getClass()));
```

Additionally, should we add properties to allow whitelisting some scenarios, such as this one?

```java
while ((result instanceof CompletionException) && (cause != null)) {
    result = cause;
    cause = result.getCause();
}
```

**Running PMD through:** *Gradle*

**Comment by jsotuyod** [comment]
@Vampire the rule will detect any pair of parenthesis that can be removed without affecting the code's behavior. As such, this being reported is effectively ok.

As for the enhancement… I guess the thing is how far we need to go… I mean, I know people tend to have trouble remembering wether `&&` or `||` takes precedence first, but how far are we gonna take it? ternaries? bit-wise operations? comparations? arithmetic operations?

I'm unsure what a sane set of properties and what scenarios they allow will be intuitive for users and reasonable to maintain.

**Comment by Vampire** [comment]
Then the rule has a false-negative instead.
In
```java
throw new IllegalArgumentException(String.format(
        "Unsupported client with class '%s' given",
        (discordClient == null ? null : discordClient.getClass())));
```
useless parantheses are found, in
```java
throw new IllegalArgumentException(String.format(
        "Unsupported client with class '%s' given",
        (discordClient == null) ? null : discordClient.getClass()));
```
they are not. 

While I also here want the parentheses to stay for easier reading.

True, finding a sane set of what to report and what not that works for all is probably impossible. :-(

**Comment by oowekyala** [comment]
For comparison, the IntelliJ inspection has a property to ignore parentheses when the inner and outer operators have different precedences. That's is not very configurable, but it's simple and makes sense.

**Comment by jsotuyod** [comment]
Ok, that one is a FP, so I'm updating the issue. I think the enhancement needs some further discussion, so maybe we should create a separate ticket.

**Comment by Vampire** [comment]
Maybe better a separate ticket for the FP to have the discussion so far here and rename this to the enhancement request?

**Comment by adangel** [comment]
Fixed with PMD 7
